src/rvsdg/egglog_optimizer/reassoc.rs

```diff
@@ -0,0 +1,74 @@
+//! This file is intended to move constants to the left
+//! for addition, promoting constant folding
+
+// TODO: we should prune eclasses with constants to just contain the constant
+// otherwise, these rules will reassociate constants many times
+// TODO: improve by also including multiplication in the normalization
+
+pub(crate) fn reassoc_rules() -> String {
+    let mut res = vec![];
+    for op in ["badd", "bmul"] {
+        // Move constants to the left using commutativity
+        res.push(format!(
+            "
+     (rule
+      ((= num (Node (PureOp (Const (IntT) (const) (Num n1)))))
+       (= lhs ({op} (IntT) other num)))
+      ((union lhs ({op} (IntT) num other))))"
+        ))
+    }
+
+    // Make all terms have the form (+ a (+ b (+ ...)))
+    // By moving nesting to the right.
+    res.push(
+        "
+      (rule
+        ((= lhs (badd (IntT)
+                      (Node (PureOp (badd (IntT) a b)))
+                      c)))
+        ((union lhs
+                (badd (IntT)
+                      a
+                      (Node (PureOp (badd (IntT) b c)))))))
+      "
+        .to_string(),
+    );
+
+    // Move constants up the tree
+    res.push(
+        "
+        (rule
+          ((= lhs (badd (IntT)
+                        a
+                        (Node (PureOp (badd (IntT) b c)))))
```

Comment:
Just a note that this rule can blow up: consider `(+ n1 (+ n2 (+ n3 (+ n4 ...))))` for numbers `n1`, ..., where there are exponential ways to reassociate. And if any number is zero, this rule cannot even saturate.

Reply:
Yep, I had a comment on that